### PR TITLE
VQ-VAE HipMRI Dataset - s4743500

### PR DESCRIPTION
**Pull request by aidanlokk**
### Generative VQ-VAE Model on 2D HipMRI Dataset

Created a pull request that adds the final implementation code for the VQ-VAE model to generate 2D HipMRI images. The README.md file provides more information about the implementation and results from training and testing the VQ-VAE model.

### Files
- `README.md`: Provides a brief overview of the project, as well as documenting results from the training and testing procedure .
- `resources/`: Figures used for the README.md file.
- `dataset.py`: Contains the dataloader for loading and preprocessing the data.
- `modules.py`: Contains the key components and modules that make up the VQ-VAE model architecture. 
- `train.py`: Contains the functionality for training and validating the model. The trained model with the best validation loss is also saved.
- `predict.py`: Contains example usage of the saved VQ-VAE model on the test dataset (unseen images)

Author: Aidan Lok, s4743500

### CODE REVIEW

**Comment by Uto0op** [comment]
Recognition problem: VQVAE on HipMRI ---- Hard Difficulty

1. Algorithm solves the problem: (5 Marks)
Reasonably clear reconstructed images, with an SSIM over 0.8;

2. Algorithm demonstrated and functions: (3 Marks)

4. Design: (1 Mark)
Highly modularized design

5. Commenting:
Sufficient docstrings and detailed comments (1 Mark)

6. Algorithm above Normal Difficulty: (5 Marks)

7. Algorithm is Hard difficulty: (5 Marks)

**Suggestions**:

- Detailed description of the model in README.md, involving brief justification on hyperparameter tuning;
- Only the reconstructed loss was plotted, without justifying the unusual trend at the beginning. The overall/combined loss is expected as a stronger indicator of monitoring model's performance ---- ideally you'd want to plot the losses the same way as you plotted the SSIM (for training and validation, respectively);
- It's good to visualize the embedding and show it capture patterns;
- Both code and text show your efforts :)

**Comment by YashMitttal** [comment]
Some additional feedback:

Overall, it is quite a good readme. You might want to add information about:

- How different choices in the encoder (e.g., kernel size, number of residual layers) might impact performance in the context of medical images.

You might want to write slightly more details in the pull request content.

Overall, good job!

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. VQVAE (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF submitted.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time submission.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-1</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_
